File: web-sources/s8-web-front/carbide/button/Button.js

import { S8WebFront } from '/s8-web-front/S8WebFront.js';
import { NeObject } from '/s8-io-bohr-neon/NeObject.js';


S8WebFront.CSS_import("/s8-web-front/carbide/button/Button.css")

const sizeMapper = function(code){
	switch(code){
		case 0x02 : return "inlined";
		case 0x04 : return "small";
		case 0x06 : return "big";
		case 0x08 : return "huge";

		default : return "big";
	}
}

const styleMapper = function(code){
	switch(code){
		case 0x02 : return "default";
		case 0x03 : return "primary";
		case 0x04 : return "outline";
		case 0x05 : return "danger";

		default : return "default";
	}
}


/**
 * Flow base feature
 */
export class Button extends NeObject {


	style = "default";


	constructor(){
		super();

		// root node
		this.node = document.createElement("button");
		this.node.innerHTML = "button";
		
		this.node.classList.add("button");

		/* <size> */
		this.size = "small";
		this.node.classList.add("button-"+this.size);
		/* </size> */


		/* <style> */
		this.style = "default";
		this.node.classList.add("button-"+this.style);
		/* </style> */

		this.isEnabled = true;
		

		/* <listeners> */

		// input		
        let _this = this;
        this.node.addEventListener("click", function (event) {
            _this.S8_vertex.runVoid("on-click");
        }, false);
		/* </listeners> */
	}


	S8_set_label(text){
		this.node.innerHTML = text;
	}
	

	/**
	 * 
	 * @param {number} code 
	 */
	S8_set_size(code){
		let assignedSize = sizeMapper(code);
		if(assignedSize != this.size){
			this.node.classList.replace("button-"+this.size, "button-"+assignedSize);
			this.size = assignedSize;
		}
	}

	
	/**
	 * 
	 * @param {number} code 
	 */
	S8_set_style(code){
		let assignedStyle = styleMapper(code);
		if(assignedStyle != this.style){
			this.node.classList.replace("button-"+this.style, "button-"+assignedStyle);
			this.style = assignedStyle;
		}
	}


	/**
	 * 
	 * @param {state} state 
	 */
	S8_set_isEnabled(state){
		if(state){
			this.enable();
		}
		else{
			this.disable();
		}
	}



	getEnvelope(){
		return this.node;
	}

	enable(){
		if(!this.isEnabled){
			this.node.removeAttribute("disabled");
			this.isEnabled = true;
		}
	}

	disable() {
		if(this.isEnabled){
			this.node.setAttribute("disabled", "");
			this.isEnabled = false;
		}
	}


	S8_render(){
		/* nothing to post-process */
	}


	S8_dispose(){
		this.node.removeEventListener("click", this.clickListener);
	}
}

File: sources/com/s8/web/front/carbide/button/Button.java
package com.s8.web.front.carbide.button;

import com.s8.io.bohr.neon.core.NeBranch;
import com.s8.io.bohr.neon.functions.none.VoidNeFunction;
import com.s8.web.front.HTML_NeNode;


/**
 * 
 * @author pierreconvert
 *
 */
public class Button extends HTML_NeNode {
	
	/**
	 * 
	 * @param branch
	 * @param label
	 * @param size
	 * @param style
	 * @return
	 */
	public static Button create(NeBranch branch, String label, Size size, Style style) {
		Button button = new Button(branch);
		
		button.setLabel(label);
		button.setSize(size);
		button.setStyle(style);
		
		return button;
	}
	
	public enum Style{
		DEFAULT(0x02), PRIMARY(0x03), OUTLINE(0x04), DANGER(0x05);
		public int code;
		private Style(int code) { this.code = code; }
	}
	
	public enum Size {
		INLINED(0x02), SMALL(0x04), BIG(0x06), HUGE(0x08);
		public int code;
		private Size(int code) { this.code = code; }
	}

	
	/**
	 * 
	 * @param branch
	 */
	public Button(NeBranch branch) {
		super(branch, "/s8-web-front/carbide/button/Button");
		
		/* to be overridden */
		vertex.setVoidMethodLambda("on-click", () -> System.out.println("I'm clicked!"));
	}
	
	
	/**
	 * 
	 * @param text
	 */
	public void setLabel(String text) {
		vertex.setStringUTF8Field("label", text);
	}
	
	
	

	/**
	 * 
	 * @param menus
	 */
	public void setSize(Size size) {
		vertex.setUInt8Field("size", size.code);
	}
	
	
	/**
	 * 
	 * @param menus
	 */
	public void setStyle(Style style) {
		vertex.setUInt8Field("style", style.code);
	}
	
	
	public void setEnabled(boolean state) {
		vertex.setBool8Field("isEnabled", state);
	}
	
	
	
	public void onClick(VoidNeFunction lambda) {
		vertex.setVoidMethod("on-click", lambda);
	}
	
}

